File: old_trunk/cost/iona/orbacus/java/main/TestIntfWChar_impl.java
// **********************************************************************
//
// Copyright (c) 2001
// IONA Technologies, Inc.
// Waltham, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

package ORBTest;

import org.omg.CORBA.*;
import org.omg.PortableServer.*;

final class TestIntfWChar_impl extends ORBTest_WChar.IntfPOA
{
    private POA m_poa;
    private char m_aWChar;
    private String m_aWString;

    public
    TestIntfWChar_impl(
	POA poa
    )
    {
	m_poa = poa;
    }

    public synchronized char
    attrWChar()
    {
	return m_aWChar;
    }

    public synchronized void
    attrWChar(
	char value
    )
    {
	m_aWChar = value;
    }

    public synchronized char
    opWChar(
	char a0,
	CharHolder a1,
	CharHolder a2
    )
    {
	m_aWChar = (char)(a0 + a1.value);
	a1.value = a2.value = m_aWChar;
	return m_aWChar;
    }

    public synchronized char
    opWCharEx(
	char a0,
	CharHolder a1,
	CharHolder a2
    )
	throws ORBTest_WChar.ExWChar
    {
	m_aWChar = (char)(a0 + a1.value);
	throw new ORBTest_WChar.ExWChar(m_aWChar);
    }

    public synchronized String
    attrWString()
    {
	return m_aWString;
    }

    public synchronized void
    attrWString(
	String value
    )
    {
	m_aWString = value;
    }

    public synchronized String
    opWString(
	String a0,
	StringHolder a1,
	StringHolder a2
    )
    {
	m_aWString = a0 + a1.value;
	a1.value = a2.value = m_aWString;
	return m_aWString;
    }

    public synchronized String
    opWStringEx(
	String a0,
	StringHolder a1,
	StringHolder a2
    )
	throws ORBTest_WChar.ExWString
    {
	m_aWString = a0 + a1.value;
	throw new ORBTest_WChar.ExWString(m_aWString);
    }

    public org.omg.PortableServer.POA
    _default_POA()
    {
	return m_poa;
    }
}

File: cost/iona/orbacus/cpp/main/TestIntfWChar_impl.cpp
// **********************************************************************
//
// Copyright (c) 2001
// IONA Technologies, Inc.
// Waltham, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

#include <includes.h>

#include <TestIntfWChar_impl.h>

#include <limits.h>

using namespace CORBA;

// ----------------------------------------------------------------------
// Helper functions for wcslen, wcscpy, wcscat. Should not be required
// with up-to-date compilers.
// ----------------------------------------------------------------------

size_t
mywcslen(const wchar_t* str)
{
    size_t len = 0;
    while(*str++ != L'\0')
        len++;

    return len;
}

wchar_t*
mywcscpy(wchar_t* dst, const wchar_t* src)
{
    wchar_t* result = dst;
    while((*dst++ = *src++))
        /*do nothing*/;

    return result;
}

wchar_t*
mywcscat(wchar_t* dst, const wchar_t* src)
{
    mywcscpy(dst + mywcslen(dst), src);

    return dst;
}


TestIntfWChar_impl::TestIntfWChar_impl()
{
}

WChar
TestIntfWChar_impl::attrWChar()
    throw (
	CORBA::SystemException
    )
{
    return m_aWChar;
}


void
TestIntfWChar_impl::attrWChar(
    WChar val
)
    throw (
	CORBA::SystemException
    )
{
    m_aWChar = val;
}

WChar
TestIntfWChar_impl::opWChar(
    WChar a0,
    WChar& a1,
    WChar& a2
)
    throw (
	CORBA::SystemException
    )
{
    m_aWChar = a0 + a1;
    a1 = a2 = m_aWChar;
    return m_aWChar;
}

WChar
TestIntfWChar_impl::opWCharEx(
    WChar a0,
    WChar& a1,
    WChar&
)
    throw (
	ORBTest_WChar::ExWChar,
	CORBA::SystemException
    )
{
    m_aWChar = a0 + a1;
    throw ORBTest_WChar::ExWChar(m_aWChar);
    return 0; // Some compilers need this
}

wchar_t*
TestIntfWChar_impl::attrWString()
    throw (
	CORBA::SystemException
    )
{
    return wstring_dup(m_aWString);
}

void
TestIntfWChar_impl::attrWString(
    const wchar_t* val
)
    throw (
	CORBA::SystemException
    )
{
    m_aWString = val;
}

wchar_t*
TestIntfWChar_impl::opWString(
    const wchar_t* a0,
    wchar_t*& a1,
    WString_out a2
)
    throw (
	CORBA::SystemException
    )
{
    m_aWString = wstring_alloc(mywcslen(a0) + mywcslen(a1));
    mywcscpy(m_aWString.inout(), a0);
    mywcscat(m_aWString.inout(), a1);
    wstring_free(a1);
    a1 = wstring_dup(m_aWString);
    a2 = wstring_dup(m_aWString);
    return wstring_dup(m_aWString);
}

wchar_t*
TestIntfWChar_impl::opWStringEx(
    const wchar_t* a0,
    wchar_t*& a1,
    WString_out
)
    throw (
	ORBTest_WChar::ExWString,
	CORBA::SystemException
    )
{
    m_aWString = wstring_alloc(mywcslen(a0) + mywcslen(a1));
    mywcscpy(m_aWString.inout(), a0);
    mywcscat(m_aWString.inout(), a1);
    throw ORBTest_WChar::ExWString(m_aWString);
    return 0; // Some compilers need this
}
